crates/core/src/host/wasmer/mod.rs

```diff
@@ -62,52 +67,90 @@ pub fn make_actor(
+        $(impl WasmPointee for $t {
+            type Pointer = u32;
+            fn write_to(self, mem: &mut MemView, ptr: Self::Pointer) -> Result<(), WasmError> {
+                let bytes = self.to_le_bytes();
+                mem.deref_slice_mut(ptr, bytes.len() as u32)?.copy_from_slice(&bytes);
+                Ok(())
+            }
+        })*
+    };
 }
+impl_pointee!(u8, u16, u32, u64);
+impl_pointee!(super::wasm_common::BufferIdx, super::wasm_common::BufferIterIdx);
+type WasmPtr<T> = <T as WasmPointee>::Pointer;
 
 /// Wraps access to WASM linear memory with some additional functionality.
-#[derive(Clone)]
+#[derive(Clone, Copy)]
 struct Mem {
     /// The underlying WASM `memory` instance.
-    pub memory: Memory,
+    pub memory: wasmtime::Memory,
 }
 
 impl Mem {
     /// Constructs an instance of `Mem` from an exports map.
-    fn extract(exports: &wasmer::Exports) -> anyhow::Result<Self> {
-        let memory = exports.get_memory("memory")?.clone();
-        Ok(Self { memory })
+    fn extract(exports: &wasmtime::Instance, store: impl wasmtime::AsContextMut) -> anyhow::Result<Self> {
+        Ok(Self {
+            memory: exports.get_memory(store, "memory").context("no memory export")?,
+        })
     }
 
     /// Creates and returns a view into the actual memory `store`.
     /// This view allows for reads and writes.
-
-    fn view<'a>(&self, store: &'a impl AsStoreRef) -> wasmer::MemoryView<'a> {
-        self.memory.view(store)
+    fn view_and_store<'a, T>(&self, store: impl Into<StoreContextMut<'a, T>>) -> (&'a mut MemView, &'a mut T) {
+        let (mem, store_data) = self.memory.data_and_store_mut(store);
+        (MemView::from_slice_mut(mem), store_data)
     }
 
-    /// Reads a slice of bytes starting from `ptr`
-    /// and lasting `len` bytes into a `Vec<u8>`.
-    ///
-    /// Returns an error if the slice length overflows a 64-bit address.
-    fn read_bytes(&self, store: &impl AsStoreRef, ptr: WasmPtr<u8>, len: u32) -> Result<Vec<u8>, MemoryAccessError> {
-        ptr.slice(&self.view(store), len)?.read_to_vec()
+    fn view_const<'a, T: 'a>(&self, store: impl Into<StoreContext<'a, T>>) -> &'a MemView {
+        MemView::from_slice(self.memory.data(store))
     }
+}
+
+#[repr(transparent)]
+struct MemView([u8]);
 
-    /// Writes `data` into the slice starting from `ptr`
-    /// and lasting `len` bytes.
-    ///
-    /// Returns an error if
-    /// - the slice length overflows a 64-bit address
-    /// - `len != data.len()`
-    fn set_bytes(
-        &self,
-        store: &impl AsStoreRef,
-        ptr: WasmPtr<u8>,
-        len: u32,
-        data: &[u8],
-    ) -> Result<(), MemoryAccessError> {
-        ptr.slice(&self.view(store), len)?.write_slice(data)
+impl MemView {
+    fn from_slice_mut(v: &mut [u8]) -> &mut Self {
+        unsafe { &mut *(v as *mut [u8] as *mut MemView) }
+    }
+    fn from_slice(v: &[u8]) -> &Self {
+        unsafe { &*(v as *const [u8] as *const MemView) }
+    }
```

Comment:
Please add `SAFETY` comments which mention that `MemView` is a transparent newtype around `[u8]`.

Reply:
Ideally I'd use something like [`bytemuck::TransparentWrapper`](https://docs.rs/bytemuck/latest/bytemuck/derive.TransparentWrapper.html) derive for those for clarity + it makes it harder to make a mistake.

Reply:
Maybe I'll go through in a separate PR and switch that manual stuff to a derive - there's at least a good few of these throughout the codebase (mostly written by me)